treewide: completely remove types.loaOf

rnhmjoj opened the pull request:
###### Motivation for this change
Finalise the removal of loaOf, started with #63103.
See also https://github.com/NixOS/nixpkgs/issues/77189#issuecomment-583927360.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

rnhmjoj committed:
treewide: completely remove types.loaOf

worldofpeace reviewed (changes requested):
Why don't we turn the warning in the type into something fatal to help people out? 
Also mention in the release notes that it is now removed entirely?

rnhmjoj committed:
nixos/doc: convert loaOf options refs to attrsOf

rnhmjoj commented:
Thank you, done. I don't know how the Nixos options integration in docbook happen, so is there a file that needs to be updated for the removal? I couldn't find any more references to `loaOf`.

Infinisil commented:
No idea either, haven't looked into it

jtojnar commented:
We replace the `?` in `id`s for `_` here:

https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/make-options-doc/options-to-docbook.xsl#L23

The id just comes from the option name itself IIRC.

rnhmjoj commented:
@jtojnar, sorry but I understand very little of what's going on in that file. I still don't see anything related to loaOf in it, what should be changed?

jtojnar commented:
We can drop the outermost `replace` if no other type adds `?` to the generated option name lika `loaOf` did in `getSubOptions`.

rnhmjoj commented:
Ok, I think I get it. It's because `loaOf` uses `<name?>` as a placeholder.
Yes, it seems no other type does this and the manual builds, so it should be safe to remove.

rnhmjoj committed:
nixos/lib/make-options-doc: remove loaOf subs

Remove the substitution for the <name?> placeholder used by loaOf,
now that the type has been deprecated.

infinisil reviewed (approved):
LGTM

cole-h commented on lib/types.nix:
```diff
@@ -326,110 +326,15 @@ rec {
+    loaOf =
+      let msg =
+        ''
+          `types.loaOf` is deprecated and mixing list with attribute values
+          is no longer possible; use only `types.attrsOf` instead.
+          See https://git.io/JUvd1 for the motivation.
```
```suggestion
          `types.loaOf` has been removed and mixing lists with attribute values
          is no longer possible; please use `types.attrsOf` instead.
          See https://git.io/JUvd1 for the motivation.
```

Even though `types.loaOf` is aliased to `types.attrsOf` now, I still think this counts as being removed. However, if this wording isn't desired, I think at least the following minor changes should be included:

```suggestion
          `types.loaOf` is deprecated and mixing lists with attribute values
          is no longer possible; please use `types.attrsOf` instead.
          See https://git.io/JUvd1 for the motivation.
```

rnhmjoj replied:
Thanks for the suggestions. The rewording looks good to me: I've implemented it.

jtojnar replied:
Could we not use URL shorteners? They obscure the target location and in this case the URL is not even that long.

rnhmjoj replied:
Switched to original link.

cole-h commented on lib/types.nix:
```diff
@@ -326,110 +326,15 @@ rec {
-        substSubModules = m: loaOf (elemType.substSubModules m);
-        functor = (defaultFunctor name) // { wrapped = elemType; };
-      };
+    # drop this in the future:
```
I think this line should stand out more (sorry for another suggestion). Maybe just add a `FIXME:` or `TODO:` prefix, so it's easier to notice that something still needs doing here?

rnhmjoj replied:
Done, I also applied the changes to `list`, which shared a similar fate.

cole-h reviewed (approved):
Diff LGTM, thanks!

rnhmjoj commented:
Thank you all!

worldofpeace commented:
Absolutely, and I return the thanks ten fold :sparkles:
